state/kafka-receiver/src/main/java/com/expediagroup/streamplatform/streamregistry/state/kafka/KafkaEventReceiver.java

```diff
@@ -0,0 +1,170 @@
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.atomic.AtomicLong;
+
+import lombok.Builder;
+import lombok.NonNull;
+import lombok.RequiredArgsConstructor;
+import lombok.Value;
+import lombok.extern.slf4j.Slf4j;
+
+import io.confluent.kafka.serializers.KafkaAvroDeserializer;
+
+import org.apache.kafka.clients.consumer.ConsumerRecord;
+import org.apache.kafka.clients.consumer.KafkaConsumer;
+import org.apache.kafka.common.TopicPartition;
+
+import com.expediagroup.streamplatform.streamregistry.state.EventCorrelator;
+import com.expediagroup.streamplatform.streamregistry.state.EventReceiver;
+import com.expediagroup.streamplatform.streamregistry.state.EventReceiverListener;
+import com.expediagroup.streamplatform.streamregistry.state.avro.AvroConverter;
+import com.expediagroup.streamplatform.streamregistry.state.avro.AvroKey;
+import com.expediagroup.streamplatform.streamregistry.state.avro.AvroValue;
+
+@Slf4j
+@RequiredArgsConstructor(access = PACKAGE)
+public class KafkaEventReceiver implements EventReceiver {
+  @NonNull private final Config config;
+  private final EventCorrelator correlator;
+  @NonNull private final AvroConverter converter;
+  @NonNull private final KafkaConsumer<AvroKey, AvroValue> consumer;
+  @NonNull private final ScheduledExecutorService executorService;
+  private volatile boolean shuttingDown = false;
+
+  public KafkaEventReceiver(Config config, EventCorrelator correlator) {
+    this(
+        config,
+        correlator,
+        new AvroConverter(),
+        new KafkaConsumer<>(consumerConfig(config)),
+        newScheduledThreadPool(2)
+    );
+  }
+
+  public KafkaEventReceiver(Config config) {
+    this(config, null);
+  }
+
+  @Override
+  public void receive(EventReceiverListener listener) {
+    executorService.execute(() -> consume(listener));
+  }
+
+  void consume(EventReceiverListener listener) {
+    var currentOffset = new AtomicLong(0L);
+    var progressLogger = executorService
+        .scheduleAtFixedRate(() -> log.info("Current offset {}", currentOffset.get()), 10, 10, SECONDS);
+
+    var topicPartition = new TopicPartition(config.getTopic(), 0);
+    var topicPartitions = List.of(topicPartition);
+
+    int partitions = consumer.partitionsFor(topicPartition.topic()).size();
+    if (partitions != 1) {
+      throw new IllegalStateException("Unsupported partition count. Require 1, got " + partitions);
+    }
+
+    long beginningOffset = consumer.beginningOffsets(topicPartitions).get(topicPartition);
+    long endOffset = consumer.endOffsets(topicPartitions).get(topicPartition);
+    log.info("Offsets: beginning[{}], end[{}]", beginningOffset, endOffset);
+
+    consumer.assign(topicPartitions);
+    consumer.seekToBeginning(topicPartitions);
+
+    var loaded = false;
+    if (endOffset == 0L) {
+      progressLogger.cancel(true);
+      log.info("Loading complete. Empty topic.");
+      listener.onEvent(LOAD_COMPLETE);
+      loaded = true;
+    }
+
```

Comment:
Maybe it is a good idea to define some lifecycle for this long running task for monitoring purposes and a more robust cancelation policy by using—for example—`RunnableFuture`

Reply:
I originally started out with a reactive streams approach.

The current approach works simply and nicely in a spring boot app as it's closeable so spring already knows how to deal with it within the application context.